### PR TITLE
Multitenancy Support using vlanid

### PR DESCRIPTION
**Pull request by tamilmani1989**
**What this PR does / why we need it**:
This PR adds multitenancy support in both cnm and cni. The user has the option to pass vlanid. If vlanid is specified, then OVS network will be setup and the packets coming out of container will be tagged with corresponding vlan id. In case of CNM, the vlanid can be specified as option in create network request. CNI will get vlanid after querying CNS for podname and namespace.

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #

**Special notes for your reviewer**:

**Release note**:

```release-note
```

### CODE REVIEW

File: network/endpoint_linux.go [hunk 1]
```diff
@@ -203,40 +323,108 @@ func (nw *network) newEndpointImpl(epInfo *EndpointInfo) (*endpoint, error) {
...
+		log.Printf("[net] Deleting ARP reply rule failed with error %v", err)
+	}
+
+	// Add MAC address translation rule.
+	log.Printf("[net] Deleting MAC DNAT rule for IP address %v and vlan %v.", ep.IPAddresses[0].IP.String(), ep.VlanID)
+	cmd = fmt.Sprintf("ovs-ofctl del-flows %s ip,nw_dst=%s,dl_vlan=%v,in_port=%s",
+		nw.extIf.BridgeName, ep.IPAddresses[0].IP.String(), ep.VlanID, ofPort)
+	_, err = common.ExecuteShellCommand(cmd)
+	if err != nil {
+		log.Printf("[net] Adding MAC DNAT rule failed with error %v", err)
```
Comment (tamilmani1989) [inline network/endpoint_linux.go]: adding - >deleting

File: network/endpoint_linux.go [hunk 2]
```diff
@@ -203,40 +323,108 @@ func (nw *network) newEndpointImpl(epInfo *EndpointInfo) (*endpoint, error) {
...
+	}
+
+	// Delete port fromk ovs bridge
+	cmd = fmt.Sprintf("ovs-vsctl del-port %v %v", nw.extIf.BridgeName, ep.HostIfName)
+	_, err = common.ExecuteShellCommand(cmd)
+	if err != nil {
+		log.Printf("[net] Adding port failed with error %v", err)
```
Comment (tamilmani1989) [inline network/endpoint_linux.go]: adding - >deleting

File: network/endpoint_linux.go [hunk 3]
```diff
@@ -25,15 +27,149 @@ const (
...
+	log.Printf("[net] Get ovs port for interface %v.", nw.extIf.Name)
+	cmd = fmt.Sprintf("ovs-vsctl get Interface %s ofport", nw.extIf.Name)
+	ofport, err := common.ExecuteShellCommand(cmd)
+	if err != nil {
+		log.Printf("[net] Get ofport failed with error %v", err)
+		return err
+	}
+
+	ofport = strings.Trim(ofport, "\n")
+
```
Comment (tamilmani1989) [inline network/endpoint_linux.go]: remove space

File: network/network_linux.go [hunk 4]
```diff
@@ -187,6 +254,182 @@ func (nm *networkManager) deleteBridgeRules(extIf *externalInterface) {
...
+	}
+
+	// Apply IP configuration to the bridge for host traffic.
+	err = nm.applyIPConfig(extIf, bridge)
+	if err != nil {
+		log.Printf("[net] Failed to apply interface IP configuration: %v.", err)
+	}
+
+	extIf.BridgeName = bridgeName
+	err = nil
```
Comment (tamilmani1989) [inline network/network_linux.go]: not needed

File: network/network_linux.go [hunk 5]
```diff
@@ -30,14 +36,32 @@ type route netlink.Route
...
+			vlanid, err = strconv.Atoi(opt[vlanIDKey].(string))
+			if err != nil {
+				log.Printf("Error while converting vlanid from string to integer")
+			}
+
+			nm.createOVSNetwork(extIf, nwInfo)
+		} else {
+			log.Printf("create linux bridge")
+			err := nm.connectExternalInterface(extIf, nwInfo)
+			if err != nil {
```
Comment (saiyan86) [inline network/network_linux.go]: let's stick with "if err := xxx; err != nil {}" format

File: network/network_linux.go [hunk 6]
```diff
@@ -56,8 +81,15 @@ func (nm *networkManager) newNetworkImpl(nwInfo *NetworkInfo, extIf *externalInt
 // DeleteNetworkImpl deletes an existing container network.
 func (nm *networkManager) deleteNetworkImpl(nw *network) error {
 	// Disconnect the interface if this was the last network using it.
-	if len(nw.extIf.Networks) == 1 {
-		nm.disconnectExternalInterface(nw.extIf)
+
+	if nw.VlanId != 0 {
+		if len(nw.extIf.Networks) == 1 {
+			nm.disconnectOVSInterface(nw.extIf)
+		}
```
Comment (saiyan86) [inline network/network_linux.go]: if we return here then the else{} is not needed anymore.

File: network/network_linux.go [hunk 7]
```diff
@@ -179,6 +211,41 @@ func (nm *networkManager) addBridgeRules(extIf *externalInterface, hostIf *net.I
...
+	ofport, err := common.ExecuteShellCommand(cmd)
+	if err != nil {
+		log.Printf("[net] Get ofport failed with error %v", err)
+		return err
+	}
+
+	ofport = strings.Trim(ofport, "\n")
+
+	// Add DNAT rule to forward ARP replies to container interfaces.
+	log.Printf("[net] Adding DNAT rule for ingress ARP traffic on interface %v.", hostIf.Name)
```
Comment (saiyan86) [inline network/network_linux.go]: I saw we have a few places using similar code blocks, can we reuse them?

File: network/network_linux.go [hunk 8]
```diff
@@ -187,6 +254,182 @@ func (nm *networkManager) deleteBridgeRules(extIf *externalInterface) {
...
+	}
+
+	// Find the external interface.
+	hostIf, err := net.InterfaceByName(extIf.Name)
+	if err != nil {
+		return err
+	}
+
+	// If a bridge name is not specified, generate one based on the external interface index.
+	bridgeName := nwInfo.BridgeName
```
Comment (saiyan86) [inline network/network_linux.go]: We can do _if bridgeName := nwInfo.BridgeName; bridgeName != ""{}_

File: network/network_linux.go [hunk 9]
```diff
@@ -187,6 +254,182 @@ func (nm *networkManager) deleteBridgeRules(extIf *externalInterface) {
...
+	}
+
+	// Check if the bridge already exists.
+	bridge, err := net.InterfaceByName(bridgeName)
+	if err != nil {
+		// Create the bridge.
+		log.Printf("[net] Creating bridge %v.", bridgeName)
+
+		err = nm.createOVSBridge(bridgeName)
+		if err != nil {
```
Comment (saiyan86) [inline network/network_linux.go]: merge with previous line, pls also fix this in other places.

File: network/network_linux.go [hunk 10]
```diff
@@ -187,6 +254,182 @@ func (nm *networkManager) deleteBridgeRules(extIf *externalInterface) {
...
+		log.Printf("[net] Failed to delete bridge %v, err:%v.", extIf.BridgeName, err)
+	}
+
+	extIf.BridgeName = ""
+
+	// Restore IP configuration.
+	hostIf, _ := net.InterfaceByName(extIf.Name)
+	err = nm.applyIPConfig(extIf, hostIf)
+	if err != nil {
+		log.Printf("[net] Failed to apply IP configuration: %v.", err)
```
Comment (saiyan86) [inline network/network_linux.go]: Should we return err here?

File: network/network_linux.go [hunk 11]
```diff
@@ -187,6 +254,182 @@ func (nm *networkManager) deleteBridgeRules(extIf *externalInterface) {
...
+
+// DisconnectExternalInterface disconnects a host interface from OVS bridge.
+func (nm *networkManager) disconnectOVSInterface(extIf *externalInterface) error {
+	log.Printf("[net] Disconnecting interface %v.", extIf.Name)
+
+	// Disconnect external interface from its bridge.
+	cmd := fmt.Sprintf("ovs-vsctl del-port %s %s", extIf.BridgeName, extIf.Name)
+	_, err := common.ExecuteShellCommand(cmd)
+	if err != nil {
+		log.Printf("[net] Failed to disconnect interface %v from bridge, err:%v.", extIf.Name, err)
```
Comment (saiyan86) [inline network/network_linux.go]: Should we return err here?

File: network/network_linux.go [hunk 12]
```diff
@@ -187,6 +254,182 @@ func (nm *networkManager) deleteBridgeRules(extIf *externalInterface) {
...
+	}
+
+	// Delete the bridge.
+	err = nm.deleteOVSBridge(extIf.BridgeName)
+	if err != nil {
+		log.Printf("[net] Failed to delete bridge %v, err:%v.", extIf.BridgeName, err)
```
Comment (saiyan86) [inline network/network_linux.go]: Should we return err here?

File: network/network_linux.go [hunk 13]
```diff
@@ -187,6 +254,182 @@ func (nm *networkManager) deleteBridgeRules(extIf *externalInterface) {
...
+	extIf.BridgeName = bridgeName
+	err = nil
+
+	log.Printf("[net] Connected interface %v to bridge %v.", extIf.Name, extIf.BridgeName)
+
+	return nil
+}
+
+// DisconnectExternalInterface disconnects a host interface from OVS bridge.
+func (nm *networkManager) disconnectOVSInterface(extIf *externalInterface) error {
```
Comment (saiyan86) [inline network/network_linux.go]: This function always return nil. Is this by design?

File: cns/service/main.go [hunk 14]
```diff
@@ -226,6 +226,31 @@ func main() {
 			return
 		}
 
+		// Set plugin options.
+		netPlugin.SetOption(acn.OptAPIServerURL, url)
+		log.Printf("Start netplugin\n")
+		err = netPlugin.Start(&pluginConfig)
+		if err != nil {
+			fmt.Printf("Failed to create network plugin, err:%v.\n", err)
```
[the remark this answers is not in the document]
Comment (tamilmani1989) [reply cns/service/main.go]: ok i moved setting up log to top and failed to change this place. will change it

File: network/network_linux.go (hunk 12 above)
Comment (tamilmani1989) [reply network/network_linux.go]: not needed. if it returns error, then we assume port already removed

File: network/network_linux.go (hunk 13 above)
Comment (tamilmani1989) [reply network/network_linux.go]: yes this was like this from before. i will remove it

File: network/network_linux.go (hunk 8 above)
Comment (tamilmani1989) [reply network/network_linux.go]: bridgename will be used elsewhere in that function. so can't do this

File: network/network_linux.go (hunk 12 above)
Comment (saiyan86) [reply network/network_linux.go]: Then we always return nil. How can caller function catch such error?

File: network/network_linux.go (hunk 7 above)
Comment (saiyan86) [reply network/network_linux.go]: stuff like get ovs port appears in several places. But this is fine, as those are not big code blocks.

File: network/network_linux.go (hunk 12 above)
Comment (tamilmani1989) [reply network/network_linux.go]: If something went wrong, we have to check logs. I didn't change any behaviour. I followed the same as that of linux bridge. If we change this, im not sure what we gonna break

**Pull request closed without merging**